storage/retrieval.go

```diff
@@ -133,69 +153,81 @@ func (b *Bucket) getLogChunks(ctx context.Context, buildID string) ([]LogChunkIn
 		}
 	}
 
-	if err := iter.Err(); err != nil {
-		return nil, nil, errors.Wrap(err, "getting log chunks")
-	}
-
 	sortLogChunksByStartTime(buildChunks)
 	sortLogChunksByStartTime(testChunks)
 
 	return buildChunks, testChunks, nil
 }
 
-// filterLogChunksByTestID returns (1) the resulting slice of log chunks after
-// filtering for chunks with the given test ID and (2) the appropriate test
-// execution time range for the given test ID.
-func filterLogChunksByTestID(chunks []LogChunkInfo, testID string) ([]LogChunkInfo, TimeRange) {
+// parseTestIDs parses test IDs from the buildKeys that correspond to test metadata files
+// and sorts them by creation time.
+func parseTestIDs(buildKeys []string) ([]model.TestID, error) {
+	var testIDs []model.TestID
+	for _, key := range buildKeys {
+		if !strings.HasSuffix(key, metadataFilename) {
+			continue
+		}
+		if !strings.Contains(key, "/tests/") {
+			continue
+		}
+		testID, err := testIDFromKey(key)
+		if err != nil {
+			return nil, errors.Wrap(err, "getting test ID from metadata key")
+		}
+		testIDs = append(testIDs, model.TestID(testID))
+	}
+
+	sort.Slice(testIDs, func(i, j int) bool {
+		return testIDs[i].Timestamp().Before(testIDs[j].Timestamp())
+	})
+
+	return testIDs, nil
+}
+
+// testExecutionWindow returns the TimeRange from the creation of this test to the creation
+// of the next test. If the testID isn't found the returned TimeRange is unbounded.
+// If there is no later test then the end time is TimeRangeMax.
+func testExecutionWindow(allTestIDs []model.TestID, testID string) TimeRange {
+	tr := AllTime
```

Comment:
This copies by value right? Because we mutate it later and that would be bad to mutate a global. We probably want to make TimeRange immutable at some point...

Reply:
I think it does copy by value.
What do you mean to make a `TimeRange` immutable?

Reply:
I don't think we need to worry about this now, but basically making the fields private so you can't change them and making public accessors. That way you don't have to worry about the whole "can someone mutate this pointer?"